### PR TITLE
Design: infer_var_type

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

**Comment by Superjom** [comment]
Similar to the previous Tensor/LoDTensor type infer? @qingqing01

**Comment by QiJune** [comment]
Maybe we should also register InferShape to OpInfo @jacquesqiao

File: doc/design/infer_var_type.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Design Doc: InferVarType
+
+## The Problem Posed
+
+The variable in our design can hold variant types. Such as `LoDTensor` and `SelectedRows`. An operator should be able to inference the variable types of its output.
+
+For example, a `lookup table` operator takes two `LoDTensor`; one is a float tensor as the embedding table, the other is an int tensor as word ID. The gradient operator of `lookup table` will generate a `SelectedRows` as its output. A `sum` operator can take both `LoDTensor` and `SelectedRows` as its inputs and will generate a `LoDTensor` if any of its inputs is `LoDTensor`, otherwise, the `sum` operator will generate `SelectedRows` as its output.
```
[the remark this answers is not in the document]
Comment (Superjom) [reply doc/design/infer_var_type.md]: Had a discussion with reyang, the new type `SelectedRow` serves in scenarios such as sum the outputs of two `lookup`. 

It seems ok and there is no better way to support this currently.

**Comment by reyoung** [comment]
> InferVarType and InferShape are actually the same thing -- it should be called "Infer Type"

@wangkuiyi 

I think there are several differences.

1. `InferVarType` will be invoked ONLY when compiling since it will not be changed at runtime. However, `InferShape` will be invoked both compile-time and run-time.
1. Most operators do not need to write `InferVarType` since they just generate `LoDTensor` as its output.

**Pull request merged**